From 467bcd2c73874675ef5e7c1ff7d2f97ca2052474 Mon Sep 17 00:00:00 2001
From: Volodymyr Mytnyk <volodymyr.mytnyk@plvision.eu>
Date: Tue, 24 Mar 2020 16:50:22 +0200
Subject: [PATCH] Add port ranges structures

ACL add support for UDP/TCP port ranges API

This is temporary solution to be able to get flow UTP/TCP port
range when offloading the ACL rule with this match. The proper
patch will be implemented separately and up-streamed into official
net-next later.

- Add patch to validation target to be able to verify
  related patch with UDP port range

---
 include/net/flow_dissector.h | 10 ++++++++++
 include/net/flow_offload.h   |  6 ++++++
 net/core/flow_offload.c      |  7 +++++++
 3 files changed, 23 insertions(+)

diff --git a/include/net/flow_dissector.h b/include/net/flow_dissector.h
index 628383915827..d4b795625124 100644
--- a/include/net/flow_dissector.h
+++ b/include/net/flow_dissector.h
@@ -162,6 +162,16 @@ struct flow_dissector_key_ports {
 	};
 };
 
+/**
+ * flow_dissector_key_ports_range:
+ *		tp_min: min port number in range
+ *		tp_max: max port number in range
+ */
+struct flow_dissector_key_ports_range {
+         struct flow_dissector_key_ports tp_min;
+         struct flow_dissector_key_ports tp_max;
+};
+
 /**
  * flow_dissector_key_icmp:
  *		type: ICMP type
diff --git a/include/net/flow_offload.h b/include/net/flow_offload.h
index c6f7bd22db60..80f492247a93 100644
--- a/include/net/flow_offload.h
+++ b/include/net/flow_offload.h
@@ -48,6 +48,10 @@ struct flow_match_ports {
 	struct flow_dissector_key_ports *key, *mask;
 };
 
+struct flow_match_ports_range {
+	struct flow_dissector_key_ports_range *key, *mask;
+};
+
 struct flow_match_icmp {
 	struct flow_dissector_key_icmp *key, *mask;
 };
@@ -90,6 +94,8 @@ void flow_rule_match_ip(const struct flow_rule *rule,
 			struct flow_match_ip *out);
 void flow_rule_match_ports(const struct flow_rule *rule,
 			   struct flow_match_ports *out);
+void flow_rule_match_ports_range(const struct flow_rule *rule,
+			   struct flow_match_ports_range *out);
 void flow_rule_match_tcp(const struct flow_rule *rule,
 			 struct flow_match_tcp *out);
 void flow_rule_match_icmp(const struct flow_rule *rule,
diff --git a/net/core/flow_offload.c b/net/core/flow_offload.c
index 45b6a59ac124..a85d70c7a413 100644
--- a/net/core/flow_offload.c
+++ b/net/core/flow_offload.c
@@ -97,6 +97,13 @@ void flow_rule_match_ports(const struct flow_rule *rule,
 }
 EXPORT_SYMBOL(flow_rule_match_ports);
 
+void flow_rule_match_ports_range(const struct flow_rule *rule,
+			   struct flow_match_ports_range *out)
+{
+	FLOW_DISSECTOR_MATCH(rule, FLOW_DISSECTOR_KEY_PORTS_RANGE, out);
+}
+EXPORT_SYMBOL(flow_rule_match_ports_range);
+
 void flow_rule_match_tcp(const struct flow_rule *rule,
 			 struct flow_match_tcp *out)
 {
-- 
2.17.1

